queryable-spec.ts

```diff
@@ -0,0 +1,424 @@
+  ): FilterableResult;
+};
+
+
+/******************************************************************************
+                              QUERYABLE SOURCE
+ *****************************************************************************/
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties. Support for
+ * immutability is required (but implementations are free to be mutable) which
+ * determines the return value of the set() and delete() methods to be an 
+ * instance of Bindings (potentially a different one).
+ */ 
+interface Bindings extends Iterable<[RDF.Variable, RDF.Term]> {
+  type: 'bindings';
+  has: (key: RDF.Variable) => boolean;
+  get: (key: RDF.Variable) => RDF.Term | undefined;
+  keys: () => Iterator<RDF.Variable>;
+  values: () => Iterator<RDF.Term>;
+  entries: () => Iterator<[RDF.Variable, RDF.Term]>;
+  forEach: (fn: (value: RDF.Term, key: RDF.Variable) => any) => void;
+  size: number;
+  [Symbol.iterator]: () => Iterator<[RDF.Variable, RDF.Term]>;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings: (entries?: [RDF.Variable, RDF.Term][]) => Bindings;
+  filter: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => boolean) => Bindings;
+  map: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => RDF.Term) => Bindings;
+  merge: (left: Bindings, right: Bindings) => Bindings;
+  mergeWith: (
+    merger: (left: RDF.Term, right: RDF.Term, key: RDF.Variable) => RDF.Term,
+    left: Bindings,
+    right: Bindings,
+  ) => Bindings;
+  set: (bindings: Bindings, key: RDF.Variable, value: RDF.Term) => Bindings;
+  delete: (bindings: Bindings, key: RDF.Variable) => Bindings;
+}
+
+/*
+ * Base interfaces to represent query results. These interfaces are generic
+ * with respect to the types of query metadata objects. These can be extended
+ * by implementors.
+ */
+
+interface QueryableResultMetadata<OrderItemsType extends TermName | RDF.Variable> {
+  cardinality?: number;
+  availableOrders?: QueryOperationOrder<OrderItemsType>[];
+  [key: string]: any;
+}
+
+interface QueryableResultBindings {
+  type: 'bindings';
+  execute(opts?: { order?: QueryOperationOrder<RDF.Variable> }): Promise<Stream<Bindings>>;
+  variables: RDF.Variable[];
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<RDF.Variable>>;
+}
+    
+interface QueryableResultQuads {
+  type: 'quads';
+  execute(opts?: { order?: QueryOperationOrder<TermName> }): Promise<Stream<RDF.Quad>>;
+  metadata(opts: { [key: string]: any }): Promise<QueryableResultMetadata<TermName>>;
+}
+
+interface QueryableResultBoolean {
+  type: 'boolean';
+  execute(): Promise<boolean>;
```

Comment:
Metadata can also be produced here (and in the void). We'lll probably need a more generic metadata supertype, as cardinality and order doesn't make sense here.